rust/lance/src/dataset/scanner.rs

```diff
@@ -522,15 +573,101 @@ impl Scanner {
             extra_columns.push(vector_field);
             extra_columns.push(ArrowField::new(DIST_COL, DataType::Float32, true));
         };
+
         if self.with_row_id {
             extra_columns.push(ROW_ID_FIELD.clone());
         }
 
         let schema = if !extra_columns.is_empty() {
-            self.projections.merge(&ArrowSchema::new(extra_columns))?
+            self.phyical_columns
+                .merge(&ArrowSchema::new(extra_columns.clone()))?
         } else {
-            self.projections.clone()
+            self.phyical_columns.clone()
         };
+
+        // drop metadata
+        // NOTE: this is the current behavior as we don't return metadata in queries
+        // We should reconsider this behavior
+        if self.nearest.is_some() {
+            Ok(Arc::new(Schema {
+                fields: schema.fields,
+                metadata: HashMap::new(),
+            }))
+        } else {
+            Ok(Arc::new(schema))
+        }
+    }
+
+    pub(crate) fn output_expr(&self) -> Result<Vec<(Arc<dyn PhysicalExpr>, String)>> {
+        let physical_schema = self.physical_schema()?;
+
+        // reset the id ordering becuase we will take the batch after projection
+        let physical_schema = ArrowSchema::new(
+            physical_schema
+                .fields
+                .iter()
+                .map(|f| ArrowField::new(f.name.clone(), f.data_type(), f.nullable))
+                .collect::<Vec<_>>(),
+        );
+        let df_schema = Arc::new(DFSchema::try_from(physical_schema.clone())?);
+
+        let project_star = self
+            .phyical_columns
+            .fields
+            .iter()
+            .map(|f| {
+                Ok((
+                    expressions::col(f.name.as_str(), &physical_schema)?.clone(),
+                    f.name.clone(),
+                ))
+            })
+            .collect::<Result<Vec<_>>>()?;
+
+        let output_expr = self.output_expr.clone().map(|exprs| {
+            exprs
+                .iter()
+                .map(|(expr, name)| {
+                    Ok((
+                        datafusion::physical_expr::create_physical_expr(
+                            expr,
+                            &df_schema,
+                            &Default::default(),
+                        )?,
+                        name.clone(),
+                    ))
+                })
+                .collect::<Result<Vec<_>>>()
+        });
+
+        // Append the extra columns
+        let mut output_expr = output_expr.unwrap_or(Ok(project_star))?;
```

Comment:
What if this function just took the existing exec node and optionally wrapped it in `DFProjectionExec`, but only if `self.output_expr` is not None? That is one way to eliminate the redundancy.

Reply:
yeah, just processing all of these in `create_plan` would make life a lot easier. but we need this a head of time to return schema. Let me make a ticket for cleaning this up.